Fix ovnkube-node startup when UDN runs out of host subnet

pperiyasamy opened the pull request:
When L3 Primary UDN host subnet is exhausted, host subnet may not be allocated for the node(s) by cluster manager, then node is not annotated with subnet on the k8s.ovn.org/node-subnets. When ovnkube-node is restarted, then it goes into CLBO because subnet is not found for the network on the annotation, but it okay UDN to fail on the node, but ovnkube-node should come up and continue to work for other networks. So this PR relaxes subnet annotation check for the UDN network on network controller's syncAll method and makes ovnkube-controller to come up without an error.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Event-driven reconciliation for User Defined Networks (UDN), improving responsiveness to node changes.
  - Cluster-aware CIDR selection with support for limited host-subnet configurations.

- Bug Fixes
  - More reliable UDN startup and reconciliation; falls back to self-recovery instead of failing.
  - Reduced unnecessary reconciliations for non-applicable topologies.
  - De-duplication and safe merging of VRF routes to avoid conflicts.
  - Consistent host-subnet manifest formatting.

- Tests
  - Expanded end-to-end coverage for UDN (limited host-subnet, EndpointSlice mirroring, preconfigured L2).
  - Manifest generation updated to honor supported network configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pperiyasamy commented:
This fix causes cluster unreachable at times when a UDN network (which doesn't have enough subnet to allocate for a node) is created/deleted multiple times. Need to investigate why that happens.

pperiyasamy commented:
fix is working fine, will add E2E to cover exhausted UDN subnet scenario.

pperiyasamy committed:
Add E2E for exhausted UDN subnet

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/retest-failed

pperiyasamy commented:
> some net-seg tests are flaking, from the initial look those are not related to PR change. running failed jobs again.

The following test is failing consistently.

```
2025-07-31T07:09:16.7229585Z   I0731 07:09:16.722420 62653 network_segmentation_services.go:442] Unexpected error: Failed to verify that Client udn-services-868/udn-client should connect to NodePort service udn-services-868/udn-service on fc00:f853:ccd:e793::4:31264 (node ovn-worker, other node): 
2025-07-31T07:09:16.7231047Z       <wait.errInterrupted>: 
2025-07-31T07:09:16.7231485Z       timed out waiting for the condition
2025-07-31T07:09:16.7231832Z       {
2025-07-31T07:09:16.7232416Z           cause: <*errors.errorString | 0xc0004843b0>{
2025-07-31T07:09:16.7233081Z               s: "timed out waiting for the condition",
2025-07-31T07:09:16.7233417Z           },
2025-07-31T07:09:16.7233587Z       }
2025-07-31T07:09:16.7234666Z   [38;5;9m[FAILED][0m in [It] - /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/network_segmentation_services.go:442 [38;5;243m@ 07/31/25 07:09:16.722[0m
2025-07-31T07:09:17.5210856Z   [1mSTEP:[0m dump namespace information after failure [38;5;243m@ 07/31/25 07:09:17.52[0m
2025-07-31T07:09:17.5212006Z   [1mSTEP:[0m Destroying namespace "udn-services-868" for this suite. [38;5;243m@ 07/31/25 07:09:17.52[0m
2025-07-31T07:09:17.5260135Z [38;5;9m• [FAILED] [54.813 seconds][0m
2025-07-31T07:09:17.5262293Z [0mNetwork Segmentation: services [38;5;243mon a user defined primary network [0mshould be reachable through their cluster IP, node port and load balancer [38;5;9m[1m[It] L3 primary UDN, cluster-networked pods, NodePort service[0m [38;5;204m[Feature:NetworkSegmentation][0m
```

https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/16618453571/job/47091502173?pr=5232

Able to repro the issue at local kind non-ic cluster. This failure happens when this test runs after ovnkube-node pod restart caused new E2E test https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5232/commits/8b34fe1c7b197da1b86b468f089ffb3ffeed702b. Still trying to figure out the root cause.

cc @ricky-rav @tssurya PTAL.

ricky-rav commented:
@pperiyasamy so are you saying that without the new e2e test you're adding, the existing e2e test on UDN services works?  And that it doesn't work especially because you restart  all ovnkube-node pods?

I'm asking because we already restart ovnkube-node, but only on the node where the client is running:
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/6082160282cf263efd162c0d04ca81ae3c9ecda7/test/e2e/network_segmentation_services.go#L262

pperiyasamy commented:
> so are you saying that without the new e2e test you're adding, the existing e2e test on UDN services works? And that it doesn't work especially because you restart all ovnkube-node pods?

@ricky-rav yes, this specific service test doesn't work only for this PR and after an ovnkube-node pod restart happened from the new E2E test.

ricky-rav commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -245,13 +245,32 @@ func (c *networkController) syncAll() error {
 	klog.Infof("%s: syncing all networks", c.name)
 	for _, network := range validNetworks {
-		err := c.syncNetwork(network.GetNetworkName())
+		// Ensure node is allocated with host subnet for the network, When subnet is not allocated
```
```suggestion
		// Ensure node is allocated with host subnet for the network. When subnet is not allocated
```

ricky-rav commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -245,13 +245,32 @@ func (c *networkController) syncAll() error {
+		// Ensure node is allocated with host subnet for the network, When subnet is not allocated
+		// for the node then skip syncing the network and proceed with syncing remaining network.
+		// Return the error at the end when k8s.ovn.org/node-subnets annotation is not set for a
+		// network and let the caller to decide on what to do with it.
```
```suggestion
		// network and let the caller decide what to do with it.
```

ricky-rav commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -245,13 +245,32 @@ func (c *networkController) syncAll() error {
+					networkName, c.node)
+				continue
+			}
+		}
```
Do you think we can move this block to a separate function, like `IsNetworkReadyForSync` or a similar name?

pperiyasamy replied:
sure, moved this new block into a `isNetworkReadyForSync`  method.

ricky-rav commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -245,13 +245,32 @@ func (c *networkController) syncAll() error {
+					c.node, networkName, err)
+			}
+			_, err = util.ParseNodeHostSubnetAnnotation(node, networkName)
+			if err != nil && util.IsAnnotationNotSetError(err) {
```
what happens if `err != nil && !util.IsAnnotationNotSetError(err)`? Should we still sync the network as in your changes or should we skip it?

pperiyasamy replied:
this PR is only handling pod crash only upon missing host subnet annotation, so it fixes a d/s issue https://issues.redhat.com/browse/OCPBUGS-55105. should we cover more cases ?

ricky-rav commented on test/e2e/network_segmentation.go:
```diff
@@ -1808,6 +1809,47 @@ spec:
 	})
+
+	Context("UDN with limited host subnet", func() {
+		It("check ovnkube-node pod restart when node is not assigned with UDN host subnet", func() {
```
```suggestion
		It("check ovnkube-node pod restart when node is not assigned a UDN host subnet", func() {
```

ricky-rav commented on test/e2e/network_segmentation.go:
```diff
@@ -1808,6 +1809,47 @@ spec:
+			netConfig := networkAttachmentConfigParams{
+				name:     nadName,
+				topology: "layer3",
+				cidr:     correctCIDRFamily(cs, "192.168.100.0/25/26", "2014:100:200::/63/64"),
```
Can you add a brief comment with the reasoning behind your e2e test? Her I understand that you're assigning a pod subnet of /25 to the cluster and of /26 to each node. So there will be only two /26 subnets available and the third node will be left without a subnet. The same goes for IPv6. Is this correct?

pperiyasamy replied:
yes @ricky-rav , you are right, added comment to reflect this.

ricky-rav commented on test/e2e/network_segmentation.go:
```diff
@@ -1808,6 +1809,47 @@ spec:
+				}
+				Expect(err).ShouldNot(HaveOccurred(), "should not be other than host subnet annotation not set error")
+			}
+			Expect(restartpod).To(BeTrue())
```
Here we're expecting to restart only 1 pod, right? So maybe `restartpod` should be an int that we expect to equal 1.

pperiyasamy replied:
there can be more than 3 nodes, so adding check now with `>= 1`.

ricky-rav reviewed (changes requested):
Overall looks good, I've left a few comments to improve clarity. Thanks, @pperiyasamy !

pperiyasamy commented:
> > > so are you saying that without the new e2e test you're adding, the existing e2e test on UDN services works? And that it doesn't work especially because you restart all ovnkube-node pods?
> > 
> > 
> > @ricky-rav yes, this specific service test doesn't work only for this PR and after an ovnkube-node pod restart happened from the new E2E test.
> 
> BTW how did you fix this?

Previously `syncAll() `throws such errors back to the caller and it was suppressed in nad controller's Start [method](https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/networkmanager/nad_controller.go#L158).
But this doesn't work and causing E2E test failure, so doing it with `syncAll()` method itself now.

ricky-rav reviewed (changes requested):
A couple more things to polish, otherwise looks good.

ricky-rav reviewed (commented):
Looks good now, thanks Peri!
I will approve it once CI is green :)

pperiyasamy commented:
@ricky-rav the CI is looking good now, the E2E needed an update for latest master (that is when CIDR contains subnet for both cluster as well as node, then filtering on supported network config won't work, so it's okay skip the filtering).

ricky-rav commented on test/e2e/network_segmentation.go:
```diff
@@ -2503,3 +2558,25 @@ func unmarshalPodAnnotationAllNetworks(annotations map[string]string) (map[strin
+}
+
+// takes ipv4 and ipv6 cidrs and returns the correct type for the cluster under test
+func selectCIDRs(cs clientset.Interface, ipv4CIDR, ipv6CIDR string) []string {
```
here you probably want to use the existing `filterCIDRs`

pperiyasamy replied:
@ricky-rav right, but `filterCIDRs` don't work when CIDR having two two slashes (/), so using these new functions to support it.

ricky-rav reviewed (dismissed):
/LGTM
Thanks, @pperiyasamy

jcaamano commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -245,19 +245,53 @@ func (c *networkController) syncAll() error {
+// isNetworkReadyForSync checks if the network is ready for sync. when a node is not allocated
+// with the host subnet for the given network, then the network can not be ready for sync, so
+// return false.
+func (c *networkController) isNetworkReadyForSync(network util.NetInfo) (bool, error) {
```
I don't think this is good for two reasons.

First, there is an assumption that `syncAll` should sync all networks or return an error (except for networks this process does not manage at all). This change does not comply with the assumption, I guess without addressing the reason the assumption is there in the first place.

Second, making this controller aware of what other controllers are doing (parsing subnet annotations?) takes it down the route of non-maintainability. This controller has a clear responsibility and we should not add more on top without proper abstractions. I would say this needs a different design at the very least.

pperiyasamy replied:
@npinaeva this is the source (https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_udn.go#L409) of the error, but if we ignore returning error from here, should we handle networks for node events ?

npinaeva replied:
hmm I am not sure why syncAll should have never existed? Every controller should have a syncAll to cleanup stale resources on restart?

jcaamano replied:
Sorry, wanted to say calling `syncNetwork` from `syncAll`, not just `syncAll`.

pperiyasamy commented on go-controller/pkg/networkmanager/network_controller.go:
```diff
@@ -50,11 +50,29 @@ func newNetworkController(name, zone, node string, cm ControllerManager, wf watc
+		nc.name,
+		nodeConfig,
+	)
+
```
right, fixed it.

pperiyasamy commented on go-controller/pkg/node/gateway_udn.go:
```diff
@@ -207,6 +207,15 @@ func (udng *UserDefinedNetworkGateway) AddNetwork() error {
+	} else if err != nil {
+		return fmt.Errorf("failed to parse k8s.ovn.org/node-subnets annotation for the network %s: %w",
+			udng.GetNetworkName(), err)
+	}
```
yes, done.

pperiyasamy commented:
@npinaeva @jcaamano updated the PR with commit [d9edfac](https://github.com/ovn-kubernetes/ovn-kubernetes/commit/d9edfac8b426ae4abd42f3991dd59f4adb7bd859), this may address your concerns. can you check ?
Meanwhile i'm investigating E2E failures for the new change.

pperiyasamy committed:
UDN Gateway: Reconcile when AddNetwork fails

Attempt to reconcile the network when plumbing fails
on the gateway side

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy committed:
Handle Node events for host subnet updates

Sync UDN gateway when node is updated with host subnet
annotation for the associated network.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
The latest PR update has UT failures, will revisit it later.
